### PR TITLE
Issue 2

### LINKED ISSUE #2: Authentication and ApiGateway

### PR DESCRIPTION
**Pull request by therrlupardo**
CHANGELOG:

- added IdentityService
    - creating new users
    - listing all users in database
    - logging in (includes generating jwt token which contains userId)
- added Common lib:
    - should include common classes between other services
    - at the moment includes only JwtUtil, which can be used to extract userId from JWT token 
- added ApiGateway:
    - acts as facade between user and services
    - secures connection with some help from IdentityService
- created simple docker-compose configuration for services above and postgres (used by IdentityService)
- added configuration for Development (local) and Production (docker) modes

### CODE REVIEW

File: server/ApiGateway/Properties/launchSettings.json [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+    "anonymousAuthentication": true,
+    "iisExpress": {
+      "applicationUrl": "http://localhost:10188",
+      "sslPort": 44375
+    }
+  },
+  "profiles": {
+    "IIS Express": {
+      "commandName": "IISExpress",
+      "launchBrowser": true,
```
Comment (karol-be) [inline server/ApiGateway/Properties/launchSettings.json]: Just out of curiosity, is this configuration used when launching the app on Docker? If so, isn't this attempt to run browser carried out?

File: server/docker-compose.yml [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+#    networks:
+#      - backend
+#    healthcheck:
+#      test: rabbitmq-diagnostics -q ping
+#      interval: 30s
+#      timeout: 30s
+#      retries: 3
+
+  postgres:
+    image: postgres:13-alpine
```
Comment (karol-be) [inline server/docker-compose.yml:45]: I heard that there was an issue with Polish diacritics when the default collation of a postgres based on alpine was there. Not that I suggest a deep dive into this right now, but if it just so happens that we'd have an issue with that, maybe someone would recall.

File: server/Common/Utils/JwtUtil.cs [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+            token = removeBearer(token);
+
+            var readToken = new JwtSecurityTokenHandler().ReadJwtToken(token);
+            var userId = ((List<Claim>) readToken.Claims).Find(c => c.Type.Equals("sub"));
+            return Guid.Parse(userId.Value);
+        }
+
+        private static string removeBearer(string token)
+        {
+            if (token.Contains("Bearer ")) token = token.Substring(7);
```
Comment (karol-be) [inline server/Common/Utils/JwtUtil.cs:26]: Do we have a convention of private method starting with small letter? I'm rather accustomed to a I think a standard convention that method names start with a capital no matter the visibility scope

File: server/Common/Utils/JwtUtil.cs [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+        public static Guid GetUserIdFromToken(string token)
+        {
+            token = removeBearer(token);
+
+            var readToken = new JwtSecurityTokenHandler().ReadJwtToken(token);
+            var userId = ((List<Claim>) readToken.Claims).Find(c => c.Type.Equals("sub"));
+            return Guid.Parse(userId.Value);
+        }
+
+        private static string removeBearer(string token)
```
Comment (karol-be) [inline server/Common/Utils/JwtUtil.cs]: Do we have a convention of naming private methods with a non-capital letter? I'm rather accustomed to what I think is a standard of naming methods with first capital letter no matter their access modifier

File: server/Common/Utils/JwtUtil.cs (hunk 3 above)
Comment (karol-be) [inline server/Common/Utils/JwtUtil.cs]: What I propose is just some syntax sugar but this just screams to me :D
```
var phrase = "Bearer ";
return token.Contains(phrase) ? token.Substring(phrase.Length) : token;
```
But this is just cosmetics so please do as you will ;)

File: server/Common/Utils/JwtUtil.cs [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+        /// Extracts user id from jwt token
+        /// </summary>
+        /// <param name="token">JWT token generated by IdentityService</param>
+        /// <returns>User identifier</returns>
+        public static Guid GetUserIdFromToken(string token)
+        {
+            token = removeBearer(token);
+
+            var readToken = new JwtSecurityTokenHandler().ReadJwtToken(token);
+            var userId = ((List<Claim>) readToken.Claims).Find(c => c.Type.Equals("sub"));
```
Comment (karol-be) [inline server/Common/Utils/JwtUtil.cs]: Not sure about this but I think we could go with `.First` here so we'd get a clear exception if something goes wrong

File: server/IdentityService/Properties/launchSettings.json [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+      "commandName": "IISExpress",
+      "launchBrowser": true,
+      "environmentVariables": {
+        "ASPNETCORE_ENVIRONMENT": "Development"
+      }
+    },
+    "IdentityService": {
+      "commandName": "Project",
+      "dotnetRunMessages": "true",
+      "launchBrowser": true,
```
Comment (karol-be) [inline server/IdentityService/Properties/launchSettings.json]: I guess these two don't really deserve to have a browser launched for them :P

File: server/docker-compose.yml (hunk 2 above)
Comment (therrlupardo) [reply server/docker-compose.yml:45]: I think we can ignore this issue at the moment. Postgres should be used only to store info about user (login, hashed password and guid), or at the moment it's supposed do only this :) If we decide to store there some more data, maybe with Polish siacritics, then we will decide what to do with that.

File: server/ApiGateway/Properties/launchSettings.json (hunk 1 above)
Comment (therrlupardo) [reply server/ApiGateway/Properties/launchSettings.json]: I am not really sure how it works, but I don't think it's used when launching on Docker (at least it didn't crash when I deleted entire file :P ). Anyway, I think that launching browser is not really necessary here

File: server/IdentityService/Properties/launchSettings.json (hunk 6 above)
Comment (therrlupardo) [reply server/IdentityService/Properties/launchSettings.json]: As above, launching browser doesn't look like crucial thing :P

File: server/Common/Utils/JwtUtil.cs (hunk 4 above)
Comment (therrlupardo) [reply server/Common/Utils/JwtUtil.cs]: Nah, just didn't notice :p I think that I have some habits from other languages :)

File: server/IdentityService/Controllers/IdentityController.cs [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+        [HttpPost]
+        [Route("login")]
+        public IActionResult Login([FromQuery] string login, [FromQuery] string password)
+        {
+            try
+            {
+                var token = _identityService.Login(login, password);
+                return new OkObjectResult(new LoginResponse(token));
+            }
+            catch (NullReferenceException e)
```
Comment (karol-be) [inline server/IdentityService/Controllers/IdentityController.cs]: Personally I think handling this by NRE could blur the situation if an exception not related to wrong credentials occurs. For example we could forget to register something in DI and quite possibly we'd get a NRE this way. Personally I'd go with something like 
``` bool TryLoginUser(login, password, out token) ```
Just like with int32.TryParse(...)

File: server/IdentityService/Models/User.cs [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+    public class User
+    {
+        public User()
+        {
+        }
+
+        public User(string login, string password)
+        {
+            Login = login;
+            PasswordHash = PasswordUtil.CreatePasswordHash(password);
```
Comment (karol-be) [inline server/IdentityService/Models/User.cs]: As PasswordUtil is static we could go further and make an internal(limited to project scope) extension method on string

File: server/IdentityService/Services/IdentityService.cs [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+            {
+                throw new NullReferenceException();
+            } 
+            return GenerateTokenFor(user);
+        }
+
+        public User CreateUser(string login, string password)
+        {
+            if (_context.Users.FirstOrDefault(u => u.Login.Equals(login)) != null)
+                throw new ArgumentException("User already exists");
```
Comment (karol-be) [inline server/IdentityService/Services/IdentityService.cs]: Could be a ```System.DuplicateNameException(); ```

File: server/IdentityService/Services/IdentityService.cs [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+        {
+            var creationTime = DateTime.Now;
+            var signingKey = new SymmetricSecurityKey(Encoding.ASCII.GetBytes(_settings.Value.Secret));
+
+            var jwt = new JwtSecurityToken(
+                _settings.Value.Iss,
+                _settings.Value.Aud,
+                CreateClaims(user, creationTime),
+                creationTime,
+                creationTime.Add(TimeSpan.FromMinutes(30)),
```
Comment (karol-be) [inline server/IdentityService/Services/IdentityService.cs:68]: I guess we could create a task in backlog to create a parametrization mechanism and use it here and there, what'cha think?

File: server/IdentityService/Services/IdentityService.cs [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+            );
+            return new JwtSecurityTokenHandler().WriteToken(jwt);
+        }
+
+        private static IEnumerable<Claim> CreateClaims(User user, DateTime time)
+        {
+            return new[]
+            {
+                new Claim(JwtRegisteredClaimNames.Sub, user.Id.ToString()),
+                new Claim(JwtRegisteredClaimNames.Jti, Guid.NewGuid().ToString()),
```
Comment (karol-be) [inline server/IdentityService/Services/IdentityService.cs:79]: Didn't know such claim existed, nice :)

File: server/IdentityService/Utils/PasswordUtil.cs [hunk 12]
```diff
@@ -0,0 +1,22 @@
...
+{
+    public static class PasswordUtil
+    {
+        public static string CreatePasswordHash(string password)
+        {
+            var sha256 = SHA256.Create();
+            return HashToString(sha256.ComputeHash(Encoding.UTF8.GetBytes(password)));
+        }
+
+        private static string HashToString(IEnumerable<byte> hash)
```
Comment (karol-be) [inline server/IdentityService/Utils/PasswordUtil.cs]: I find 
``` 
private static string HashToString(IEnumerable<byte> hash) =>
            hash
                .Aggregate(new StringBuilder(), (sb, b) => sb.Append(b.ToString("x2")))
                .ToString();
```
neat :D

File: server/IdentityService/Dockerfile [hunk 13]
```diff
@@ -0,0 +1,21 @@
+FROM mcr.microsoft.com/dotnet/aspnet:5.0-buster-slim AS base
+WORKDIR /app
+EXPOSE 80
+EXPOSE 443
+
+FROM mcr.microsoft.com/dotnet/sdk:5.0-buster-slim AS build
+WORKDIR /src
+COPY ["IdentityService/IdentityService.csproj", "IdentityService/"]
+RUN dotnet restore "IdentityService/IdentityService.csproj"
+COPY . .
```
Comment (karol-be) [inline server/IdentityService/Dockerfile:10]: Genuinely curious what it does :D

File: server/IdentityService/Dockerfile [hunk 14]
```diff
@@ -0,0 +1,21 @@
...
+WORKDIR "/src/IdentityService"
+RUN dotnet build "IdentityService.csproj" -c Release -o /app/build
+
+FROM build AS publish
+RUN dotnet publish "IdentityService.csproj" -c Release -o /app/publish
+
+FROM base AS final
+WORKDIR /app
+COPY --from=publish /app/publish .
+HEALTHCHECK --interval=10s --timeout=10s --retries=3 CMD curl --fail http://localhost:80/healthcheck || exit 1
```
Comment (karol-be) [inline server/IdentityService/Dockerfile:20]: This healthcheck endpoint comes bundled in with ASP :O?

File: server/ApiGateway/Startup.cs [hunk 15]
```diff
@@ -0,0 +1,76 @@
...
+                ClockSkew = TimeSpan.Zero,
+                RequireExpirationTime = true
+            };
+
+            services.AddAuthentication()
+                .AddJwtBearer("TestKey", x =>
+                {
+                    x.RequireHttpsMetadata = false;
+                    x.TokenValidationParameters = tokenValidationParameters;
+                });
```
Comment (karol-be) [inline server/ApiGateway/Startup.cs]: Purely cosmetic - nice touch is to move code like lines 32-51 here to an extension method like `void AddRSWWApiGatewayAuthentication(this IServicesCollection...)`

**Review by karol-be — APPROVED** [review]

File: server/IdentityService/Utils/PasswordUtil.cs (hunk 12 above)
Comment (therrlupardo) [reply server/IdentityService/Utils/PasswordUtil.cs]: Correct me if I do not understand it correctly: in Aggregate first parameter is base value, second is function, which is executed for every element of enumarable?

File: server/IdentityService/Services/IdentityService.cs (hunk 10 above)
Comment (therrlupardo) [reply server/IdentityService/Services/IdentityService.cs:68]: We can do it, but at the moment I don't see any other example where we can use it (or at least on current state of project :p)

File: server/IdentityService/Utils/PasswordUtil.cs (hunk 12 above)
Comment (karol-be) [reply server/IdentityService/Utils/PasswordUtil.cs]: Correct :)

**Review by karol-be — APPROVED** [review]
Nice!

**Pull request merged**